src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostTests.cs

```diff
@@ -36,6 +36,76 @@ public async Task StopAsyncWithCancellation()
             await host.StopAsync(cts.Token);
         }
 
+        public static IEnumerable<object[]> StartAsync_StopAsync_Concurrency_TestCases
+        {
+            get
+            {
+                foreach (bool stopConcurrently in new[] { true, false })
+                {
+                    foreach (bool startConcurrently in new[] { true, false })
+                    {
+                        foreach (int hostedServiceCount in new[] { 1, 4, 10 })
+                        {
+                            yield return new object[] { stopConcurrently, startConcurrently, hostedServiceCount };
+                        }
+                    }
+                }
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(StartAsync_StopAsync_Concurrency_TestCases))]
+        public async Task StartAsync_StopAsync_Concurrency(bool stopConcurrently, bool startConcurrently, int hostedServiceCount)
+        {
+            var hostedServices = new DelegateHostedService[hostedServiceCount];
+            bool[,] events = new bool[hostedServiceCount, 2];
+
+            for (int i = 0; i < hostedServiceCount; i++)
+            {
+                var service = new DelegateHostedService(() => { events[i, 0] = true; }, () => { events[i, 1] = true; } , () => { });
+
+                hostedServices[i] = service;
```

Comment:
Oops, now I see why the index were cached
```suggestion
                var index = i;
                var service = new DelegateHostedService(() => { events[index, 0] = true; }, () => { events[index, 1] = true; } , () => { });

                hostedServices[index] = service;
```